### PR TITLE
feat: Add Caching

### PR DESCRIPTION
**Pull request by zhenlei520**
# Description

Add Caching

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: src/Contrib/Caching/Masa.Contrib.Caching.MultilevelCache/README.zh-CN.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+```C#
+builder.Services.AddStackExchangeRedisCache(new RedisConfigurationOptions()
+{
+    DefaultDatabase = 1,
+    ConnectionPoolSize = 10,
+    Servers = new List<RedisServerOptions>()
+    {
+        new("localhost", 6379)
+    }
+}).AddMultilevelCache(new MultilevelCacheOptions()
```
Comment (doddgu) [inline src/Contrib/Caching/Masa.Contrib.Caching.MultilevelCache/README.zh-CN.md:72]: code format.

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/BuildingBlocks/Caching/Masa.BuildingBlocks.Caching/AbstractDistributedCacheClient.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace Masa.BuildingBlocks.Caching;
+
+public abstract class AbstractDistributedCacheClient : AbstractCacheClient, IDistributedCacheClient
+{
+    public abstract T? GetOrSet<T>(string key, Func<CacheEntry<T>> setter);
+
+    public abstract Task<T?> GetOrSetAsync<T>(string key, Func<CacheEntry<T>> setter);
+
+    public void Refresh(params string[] keys) => Refresh(GetKeys(keys));
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/BuildingBlocks/Caching/Masa.BuildingBlocks.Caching/AbstractDistributedCacheClient.cs:12]: Lifecycle of flushing cache

File: src/Contrib/Caching/Distributed/Masa.Contrib.Caching.Distributed.StackExchangeRedis/README.zh-CN.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    "DefaultDatabase": 3,
+    "ConnectionPoolSize": 10
+  }
+}
+```
+
+2. 添加Redis缓存
+
+```C#
+builder.Services.AddStackExchangeRedisCache();
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Contrib/Caching/Distributed/Masa.Contrib.Caching.Distributed.StackExchangeRedis/README.zh-CN.md:38]: in AppSettings.json

File: src/Contrib/Authentication/OpenIdConnect/Masa.Contrib.Authentication.OpenIdConnect.Cache/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -15,10 +15,10 @@ public static IServiceCollection AddOidcCache(this IServiceCollection services,
 
     public static IServiceCollection AddOidcCache(this IServiceCollection services, RedisConfigurationOptions options)
     {
-        services.AddMasaRedisCache(Constants.DEFAULT_CLIENT_NAME, options).AddMasaMemoryCache(options =>
+        services.AddStackExchangeRedisCache(Constants.DEFAULT_CLIENT_NAME, options).AddMultilevelCache(new MultilevelCacheOptions()
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Contrib/Authentication/OpenIdConnect/Masa.Contrib.Authentication.OpenIdConnect.Cache/ServiceCollectionExtensions.cs:18]: This one is waiting to be voted on, it has not yet been decided

File: src/BuildingBlocks/Caching/Masa.BuildingBlocks.Caching/Extensions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Microsoft.Extensions.DependencyInjection;
+
+public static class ServiceCollectionExtensions
+{
+    public static IServiceCollection AddDistributedCacheCore(this IServiceCollection services)
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/BuildingBlocks/Caching/Masa.BuildingBlocks.Caching/Extensions/ServiceCollectionExtensions.cs:8]: The role here is to add the default distributed cache factory as well as the default distributed cache, any good names?

File: src/Contrib/Configuration/ConfigurationApi/Masa.Contrib.Configuration.ConfigurationApi.Dcc/Internal/Extensions/DistributedCacheClientExtensions.cs [hunk 6]
```diff
@@ -15,7 +15,7 @@ internal static class DistributedCacheClientExtensions
 
         string partialKey =
             $"{environment}-{cluster}-{appId}".ToLower();
-        List<string> keys = distributedCacheClient.GetKeys($"{partialKey}*");
+        List<string> keys = distributedCacheClient.GetKeys($"{partialKey}*").ToList();
```
Comment (doddgu) [inline src/Contrib/Configuration/ConfigurationApi/Masa.Contrib.Configuration.ConfigurationApi.Dcc/Internal/Extensions/DistributedCacheClientExtensions.cs:18]: I don't think it need to `ToList`.

**Review by doddgu — CHANGES_REQUESTED** [review]

**Pull request merged**